### ISSUE TITLE
Left-shift of 0 yields negative number

### ISSUE THREAD

**Issue by murilogiacometti** [ocaml/ocaml#8864]
### Compilation

```
ocamlopt -O3 -w -5-26 error.ml
```

```
ocamlc -w -5-26 error.ml
```

### Program (error.ml)
```
let (i:int) = (lsl) (compare 0 0) 63 in print_int i
```

### Output - Bytecode (ocamlc)

```
0
```

### Output - Native (ocamlopt)

```
-4611686018427387904
```

### Environment

```
environment       ocaml-base-compiler.4.08.0
ocamlc            4.08.0
ocamlopt          4.08.0
architecture/OS   x86_64 GNU/Linux
```

Issue found with https://github.com/jmid/efftester

**Comment by hcarty** [comment]
I get the same result with OCaml 4.07.1 on macos.

If `compare 0 0` is replaced by a literal `0` then the output is `0` when using `ocamlopt`.

**Comment by lthls** [comment]
I can reproduce, the bug has been here since the merge of #17.
The problem is that `lsl (-1) 63` gives `0` (both on bytecode and native, as expected) but `Misc.no_overflow_lsl (-1) 63` returns `true`.
I'll submit a fix.

**Comment by lthls** [comment]
@murilogiacometti could you precise what setup of the compiler you're testing ?
It is possible that your compiler is compiled without flambda support, which means that the `-O3` option is ignored. If you have the opportunity, it could be interesting to compare not only `ocamlc` and `ocamlopt` but also an flambda-enabled `ocamlopt` (to which you could pass the `-O3` option if you wish).
Thanks for the report anyway.

**Comment by hcarty** [comment]
FWIW, my test on macos was with a default/non-flambda `ocamlopt`.

**Comment by lthls** [comment]
(Both flambda and non-flambda are affected, but since this bug was found with the help of an automated tool, testing more configurations could find more bugs)

**Comment by xavierleroy** [comment]
Fixed by #8865.  Commits: trunk 0557f07, 4.09 bd31ecb.
